---
feature_text: |
  ## Precision Medicine
title: RNAseq Fusion Calling
categories:
    - Module-06-RNAseq
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0006-05-01
---

**Dev notes:**
- Consider providing built kallisto index in data
- Consider providing fastqs for just chr 6 & 17. Bam splitting below takes on the order of hours
- loop or script bam to fastq process
- pizzly takes between 30 minutes and 1 hour for 2-chromosome setup

---

# Introduction

In addition to providing information about gene expression, RNA-seq data can be used to discover transcripts which result from chromosomal translocations. Translocations and their resultant chimeric (AKA fusion) transcripts are important driver mutations in many cancers. A [variety of specialized RNA alignment and filtering strategies](https://www.ncbi.nlm.nih.gov/pubmed/27485475) have been developed to identify fusion transcripts, but these programs suffer from low specificity (i.e. many false-positives) and poor correlation across programs.

This tutorial uses the [kallisto](https://pachterlab.github.io/kallisto/about) and [pizzly](https://github.com/pmelsted/pizzly) tools for fusion detection from RNA-seq data. Kallisto quantifies transcript abundance through [pseudoalignment](http://tinyheero.github.io/2015/09/02/pseudoalignments-kallisto.html). Pizzly aligns reads which kallisto has flagged as potentially spanning fusion junctions. In addition to RNA fastq files, fusion calling requires a reference transcriptome and gene annotation file- see below.

# Setup

**Prerequisites- This module assumes you have completed prior modules
including:**
- From [Installation](http://pmbio.org/module-01-setup/0001/04/01/Software_Installation/), installed Conda package manager, conda-forge and bioconda channels, samtools, sambamba, pizzly, and kallisto.
- From [RNAseq Expression Estimation](http://pmbio.org/module-06-rnaseq/0006/02/01/RNAseq_Expression/), have alignments of normal and tumor RNA as .bam files. 

**Additional setup:**

**_Important: pizzly will not work with the most recent Ensembl human GTF annotation file. Download the version 87 GTF as shown in the below code block._**

- Download Ensembl GTF and cDNA files:

```bash
cd /data/reference
wget ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.gtf.gz
wget ftp://ftp.ensembl.org/pub/release-93/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz
```

- Fusion alignment can be computationally demanding. To complete the exercise in a reasonable amount of time, we will perform fusion alignment only for chromosomes 6 and 17. We can separate original fastq reads from those chromosomes using the alignment bam files generated by HISAT2 earlier in this module:

ADD NOTE FOR FULL VERSION

```bash
# Slice
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395BL_4-chr6.bam /data/alignment/huiming/normal/HCC1395BL_RNA_H3MYFBBXX_4_CTTGTA.bam chr6
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395BL_5-chr6.bam /data/alignment/huiming/normal/HCC1395BL_RNA_H3MYFBBXX_5_CTTGTA.bam chr6
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395BL_4-chr17.bam /data/alignment/huiming/normal/HCC1395BL_RNA_H3MYFBBXX_4_CTTGTA.bam chr17
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395BL_5-chr17.bam /data/alignment/huiming/normal/HCC1395BL_RNA_H3MYFBBXX_5_CTTGTA.bam chr17
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395_4-chr6.bam /data/alignment/huiming/tumor/HCC1395_RNA_H3MYFBBXX_4_GCCAAT.bam chr6
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395_5-chr6.bam /data/alignment/huiming/tumor/HCC1395_RNA_H3MYFBBXX_5_GCCAAT.bam chr6
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395_4-chr17.bam /data/alignment/huiming/tumor/HCC1395_RNA_H3MYFBBXX_4_GCCAAT.bam chr17
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/HCC1395_5-chr17.bam /data/alignment/huiming/tumor/HCC1395_RNA_H3MYFBBXX_5_GCCAAT.bam chr17

# Sort
mkdir /data/tmp
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395_4-chr17.bam
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395_4-chr6.bam
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395_5-chr17.bam
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395_5-chr6.bam
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395BL_4-chr17.bam
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395BL_4-chr6.bam
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395BL_5-chr17.bam
~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 /data/RNA_seq/alignments/HCC1395BL_5-chr6.bam

# Merge
~/workspace/bin/sambamba merge -t 4 /data/RNA_seq/alignments/tumor-6-17.bam /data/RNA_seq/alignments/HCC1395_*sorted.bam
~/workspace/bin/sambamba merge -t 4 /data/RNA_seq/alignments/normal-6-17.bam /data/RNA_seq/alignments/HCC1395BL*sorted.bam

# Bam to fastq
~/workspace/bin/samtools bam2fq /data/RNA_seq/alignments/normal-6-17.bam > /data/RNA_seq/fastqs_RNA/normal-R-chr6-17.fastq.gz
~/workspace/bin/samtools bam2fq /data/RNA_seq/alignments/tumor-6-17.bam > /data/RNA_seq/fastqs_RNA/tumor-R-ch6-17.fastq.gz

# Split fastqs
cat /data/RNA_seq/fastqs_RNA/normal-R-chr6-17.fastq.gz | grep '^@.*/1$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/normal-R1-chr6-17.fastq.gz
cat /data/RNA_seq/fastqs_RNA/normal-R-chr6-17.fastq.gz | grep '^@.*/2$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/normal-R2-chr6-17.fastq.gz
cat /data/RNA_seq/fastqs_RNA/tumor-R-chr6-17.fastq.gz | grep '^@.*/1$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/tumor-R1-chr6-17.fastq.gz
cat /data/RNA_seq/fastqs_RNA/tumor-R-chr6-17.fastq.gz | grep '^@.*/2$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/tumor-R2-chr6-17.fastq.gz
```

# Run Fusion Alignment 
- Create kallisto index:

```bash
mkdir /data/RNA_seq/fusion
cd /data/RNA_seq/fusion
kallisto index -i index.idx -k 31 ~/data/reference/Homo_sapiens.GRCh38.cdna.all.fa.gz
```

- Quantify potential fusions:

```bash
cd /data/RNA_seq/fusion
~/workspace/bin/kallisto quant -i index.idx --fusion -o output-normal /data/RNA_seq/fastqs_RNA/normal-R1-chr6-17.fastq.gz /data/RNA_seq/fastqs_RNA/normal-R2-chr6-17.fastq.gz
~/workspace/bin/kallisto quant -i index.idx --fusion -o output-tumor /data/RNA_seq/fastqs_RNA/tumor-R1-chr6-17.fastq.gz /data/RNA_seq/fastqs_RNA/tumor-R2-chr6-17.fastq.gz
```

- Call fusions with pizzly:

```bash
cd /data/RNA_seq/fusion
~/workspace/bin/pizzly -k 31 --gtf /data/annotation/Homo_sapiens.GRCh38.87.gtf.gz --cache index.cache.txt --align-score 2 --insert-size 400 --fasta /data/annotation/Homo_sapiens.GRCh38.cdna.all.fa.gz --output normal output-normal/fusion.txt
~/workspace/bin/pizzly -k 31 --gtf /data/annotation/Homo_sapiens.GRCh38.87.gtf.gz --cache index.cache.txt --align-score 2 --insert-size 400 --fasta /data/annotation/Homo_sapiens.GRCh38.cdna.all.fa.gz --output tumor output-tumor/fusion.txt
```


