---
feature_text: |
  ## Precision Medicine
title: RNAseq Fusion Calling
categories:
    - Module-06-RNAseq
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0006-05-01
---

# Introduction

In addition to providing information about gene expression, RNA-seq data can be used to discover transcripts which result from chromosomal translocations. Translocations and their resultant chimeric (AKA fusion) transcripts are important driver mutations in many cancers. A [variety of specialized alignment and filtering strategies](https://www.ncbi.nlm.nih.gov/pubmed/27485475) have been developed to identify fusion transcripts from RNA, but these programs suffer from low specificity (i.e. many false-positives) and poor correlation across methods.

This tutorial uses the [kallisto](https://pachterlab.github.io/kallisto/about) and [pizzly](https://github.com/pmelsted/pizzly) tools for fusion detection from RNA-seq data. Kallisto quantifies transcript abundance through [pseudoalignment](http://tinyheero.github.io/2015/09/02/pseudoalignments-kallisto.html). Pizzly aligns reads which kallisto has flagged as potentially spanning fusion junctions. In addition to RNA fastq files, fusion calling requires a reference transcriptome and gene annotation file- see below.

# Setup


**Prerequisites- This module assumes you have completed prior modules
including:**
- From [Installation](http://pmbio.org/module-01-setup/0001/04/01/Software_Installation/), installed Conda package manager, conda-forge and bioconda channels, samtools, sambamba, pizzly, and kallisto.
- From [RNAseq Expression Estimation](http://pmbio.org/module-06-rnaseq/0006/02/01/RNAseq_Expression/), have alignments of normal and tumor RNA as .bam files. 

**Additional setup:**

**_Important: pizzly will not work with the most recent Ensembl human GTF annotation file. Download the version 87 GTF as shown in the below code block._**

- Download Ensembl GTF and cDNA files:

```bash
cd /data/reference
wget ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.gtf.gz
gunzip -k Homo_sapiens.GRCh38.87.gtf.gz
wget ftp://ftp.ensembl.org/pub/release-93/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz
```

- Fusion alignment can be computationally demanding. To complete the exercise in a reasonable amount of time, we will perform fusion alignment only for a small portion of two chromosomes, parts of the short arms of chromosome 6 and 17. We can separate original fastq reads from those chromosomes using the alignment bam files generated by HISAT2 earlier in this module:

```bash
# Slice merged alignment bams from HISAT2 alignment: 
mkdir /data/RNA_seq/alignments/sliced
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/sliced/HCC1395BL-6p.bam /data/RNA_seq/alignments/HCC1395BL.bam chr6:1-57200000
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/sliced/HCC1395BL-17p.bam /data/RNA_seq/alignments/HCC1395BL.bam chr17:1-22700000
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/sliced/HCC1395-6p.bam /data/RNA_seq/alignments/HCC1395.bam chr6:1-57200000
~/workspace/bin/sambamba slice -o /data/RNA_seq/alignments/sliced/HCC1395-17p.bam /data/RNA_seq/alignments/HCC1395.bam chr17:1-22700000

# Sort
mkdir /data/tmp
for f in /data/RNA_seq/alignments/sliced/*; do unsorted+=("$f"); done
for f in "${unsorted[@]}"; do ~/workspace/bin/sambamba sort --tmpdir /data/tmp -t 4 $f; done

# Merge
~/workspace/bin/sambamba merge -t 4 /data/RNA_seq/alignments/normal-6-17.bam /data/RNA_seq/alignments/sliced/HCC1395BL-6p.sorted.bam /data/RNA_seq/alignments/sliced/HCC1395BL-17p.sorted.bam
~/workspace/bin/sambamba merge -t 4 /data/RNA_seq/alignments/tumnor-6-17.bam /data/RNA_seq/alignments/sliced/HCC1395-6p.sorted.bam /data/RNA_seq/alignments/sliced/HCC1395-17p.sorted.bam

# Bam to fastq
~/workspace/bin/samtools bam2fq /data/RNA_seq/alignments/normal-6-17.bam > /data/RNA_seq/fastqs_RNA/normal-R-chr6-17.fastq.gz
~/workspace/bin/samtools bam2fq /data/RNA_seq/alignments/tumnor-6-17.bam > /data/RNA_seq/fastqs_RNA/tumor-R-chr6-17.fastq.gz

# Split fastqs
cat /data/RNA_seq/fastqs_RNA/normal-R-chr6-17.fastq.gz | grep '^@.*/1$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/normal-R1-chr6-17.fastq.gz
cat /data/RNA_seq/fastqs_RNA/normal-R-chr6-17.fastq.gz | grep '^@.*/2$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/normal-R2-chr6-17.fastq.gz
cat /data/RNA_seq/fastqs_RNA/tumor-R-chr6-17.fastq.gz | grep '^@.*/1$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/tumor-R1-chr6-17.fastq.gz
cat /data/RNA_seq/fastqs_RNA/tumor-R-chr6-17.fastq.gz | grep '^@.*/2$' -A 3 --no-group-separator > /data/RNA_seq/fastqs_RNA/tumor-R2-chr6-17.fastq.gz
```

# Run Fusion Alignment 
- Create kallisto index:

```bash
mkdir /data/RNA_seq/fusion
cd /data/RNA_seq/fusion
~/workspace/bin/kallisto index -i index.idx -k 31 ~/data/reference/Homo_sapiens.GRCh38.cdna.all.fa.gz
```

- Quantify potential fusions:

```bash
cd /data/RNA_seq/fusion
~/workspace/bin/kallisto quant -i index.idx --fusion -o kquant-normal-617 /data/RNA_seq/fastqs_RNA/normal-R1-chr6-17.fastq.gz /data/RNA_seq/fastqs_RNA/normal-R2-chr6-17.fastq.gz
~/workspace/bin/kallisto quant -i index.idx --fusion -o kquat-tumor-617 /data/RNA_seq/fastqs_RNA/tumor-R1-chr6-17.fastq.gz /data/RNA_seq/fastqs_RNA/tumor-R2-chr6-17.fastq.gz
```

- Call fusions with pizzly:

```bash
# output is created in the directory where command is run
cd /data/RNA_seq/fusion
~/workspace/bin/pizzly -k 31 --gtf /data/annotation/Homo_sapiens.GRCh38.87.gtf.gz --cache index-normal-617.cache.txt --align-score 2 --insert-size 400 --fasta /data/annotation/Homo_sapiens.GRCh38.cdna.all.fa.gz --output normal617 kquant-normal-617/fusion.txt
~/workspace/bin/pizzly -k 31 --gtf /data/annotation/Homo_sapiens.GRCh38.87.gtf.gz --cache index.cache.txt --align-score 2 --insert-size 400 --fasta /data/annotation/Homo_sapiens.GRCh38.cdna.all.fa.gz --output tumor617 kquant-tumor/fusion.txt
```

# For full RNA seq data:

```bash
# Merge fastq files
cat /data/RNA_seq/fastqs_RNA/*Norm*_R1* >  /data/RNA_seq/fastqs_RNA/normal-RNA-R1.fastq.gz
cat /data/RNA_seq/fastqs_RNA/*Norm*_R2* >  /data/RNA_seq/fastqs_RNA/normal-RNA-R2.fastq.gz
cat /data/RNA_seq/fastqs_RNA/*Tumor*_R1* >  /data/RNA_seq/fastqs_RNA/tumor-RNA-R1.fastq.gz
cat /data/RNA_seq/fastqs_RNA/*Tumor*_R2* >  /data/RNA_seq/fastqs_RNA/tumor-RNA-R2.fastq.gz

# Create kallisto index
~/workspace/bin/kallisto index -i /data/RNA_seq/fusion/index.idx -k 31 ~/data/reference/Homo_sapiens.GRCh38.cdna.all.fa.gz

# Quantify potential fusions
~/workspace/bin/kallisto quant -i /data/RNA_seq/fusion/index.idx --fusion -o /data/RNA_seq/fusion/normal-quant /data/RNA_seq/fastqs_RNA/normal-RNA-R1.fastq.gz /data/RNA_seq/fastqs_RNA/normal-RNA-R2.fastq.gz
~/workspace/bin/kallisto quant -i /data/RNA_seq/fusion/index.idx --fusion -o /data/RNA_seq/fusion/tumor-quant /data/RNA_seq/fastqs_RNA/tumor-RNA-R1.fastq.gz /data/RNA_seq/fastqs_RNA/tumor-RNA-R2.fastq.gz

# Call fusions with pizzly
# (use unzipped GTF file)
pizzly -k 31 -gtf /data/reference/Homo_sapiens.GRCh38.87.gtf --cache /data/RNA_seq/fusion/normal-index.cache.txt --align-score 2 --insert-size 400 --fasta /data/reference/Homo_sapiens.GRCh38.cdna.all.fa.gz --output /data/RNA_seq/fusion/normal-fusion /data/RNA_seq/fusion/normal-quant/fusion.txt 

pizzly -k 31 -gtf /data/reference/Homo_sapiens.GRCh38.87.gtf --cache /data/RNA_seq/fusion/tumor-index.cache.txt --align-score 2 --insert-size 400 --fasta /data/reference/Homo_sapiens.GRCh38.cdna.all.fa.gz --output /data/RNA_seq/fusion/tumor-fusion /data/RNA_seq/fusion/tumor-quant/fusion.txt 
```

