---
feature_text: |
  ## Precision Medicine
title: Somatic Variant Interpretation
categories:
    - Module-07-Clinical
feature_image: "assets/genvis-dna-bg_optimized_v1a.png"
date: 0007-03-01
---

Introduction to DGIdb, CRAVAT, and CIViC for Clinical Interpretation of Variants in Cancer

There are a large number of tools and online resources that we use to help interpret cancer variants. We will explore just a few of these in this section.  A more comprehensive list of those that we commonly use is provided below. 

### Exercise
Start with our final list of somatic variants and select a priority set.  For example, start with the variants here:
* /workspace/somatic/final/

Many of the resource below could be helpful in this exercise. 

#### Useful tools/resources for somatic cancer variant interpretation
The following tools are generally applicable to understanding cancer variants. There are hundreds of such tools.  These are ones we particularly recommend:
* CIViC - A resource for Clinical Interpreation of Variants in Cancer (WASHU). Based on expert curation of the literature.
* OncoKB - Another resource for Clinical Interpretation of Variant in Cancer (MSKCC). Based on expert curation of the literature.
* VICC knowledgebase aggregator - A search engine and normalization approach/database that pulls together knowledge from several resources including CIViC and OncoKB.
* DGIdb - A drug gene interaction database. Can be used to identify inhibitors of activated genes.
* CRAVAT - A resource for functional and oncogenic analysis of somatic variants.
* CBioPortal - Analysis of large cohorts of cancer data including the Cancer Genome Atlas (TCGA).
* ICGC - A data portal summarizing results from the International Cancer Genome Consortium
* GDC - A data portal allowing access to the harmonized analysis results for the TCGA.
* COSMIC - Probably the largest collection of observations of somatic mutations in tumors.
* Cancer Gene Census - A curate list of cancer relevant genes.
* ProteinPaint - A visualization interface for placing an observed amino acid change in the context of protein domains and mutation hotspots according to COSMIC, as well as ClinVar observations.

